lib/instrumentation/modules/mongodb.js

```diff
@@ -175,13 +173,49 @@ module.exports = (mongodb, agent, { version, enabled }) => {
     }
   }
 
-  function onEnd(event) {
+  function onSuccess(event) {
+    // `event` is a `CommandSucceededEvent`
+    // https://github.com/mongodb/specifications/blob/master/source/command-logging-and-monitoring/command-logging-and-monitoring.rst#command-succeeded-message
+    // E.g. with mongodb@6.2.0:
+    //   CommandSucceededEvent {
+    //     connectionId: '127.0.0.1:27017',
+    //     requestId: 1,
+    //     duration: 1,
+    //     reply: { ... }
+    //   }
     if (!activeSpans.has(event.requestId)) return;
-    const { failure, reply } = event;
-    const hasFailed = failure || (reply && reply.writeErrors);
+
+    const span = activeSpans.get(event.requestId);
+    activeSpans.delete(event.requestId);
+
+    // From mongodb@3.6.0 and up some commands like `deleteOne` may contain
+    // error data inside the `reply` property. It makes sense to capture it.
+    const writeErrors = event && event.reply && event.reply.writeErrors;
```

Comment:
FWIW, I think we could use optional chaining with our new min-node version now: https://node.green/#ES2020-features-optional-chaining-operator-----
No need to. Just saying it is an option for our new code now.

Reply:
IMO this a proper use case for optional chaining. I'll add it in the next commit. Thanks for pointing that out :)